drake/core/Function.h

```diff
@@ -0,0 +1,148 @@
+    bool isa(Form base) { return isa(form, base); }
+
+    static Form leastCommonAncestor(const Form &f1, const Form &f2) {
+      if (f1 == ARBITRARY || f2 == ARBITRARY) return ARBITRARY;
+      if (isa(f2, f1)) return f1;
+      return leastCommonAncestor(derivesFrom(f1), f2);
+    }
+
+    static Form leastCommonAncestor(std::initializer_list<Form> forms) {
+      if (forms.size() < 1) throw std::runtime_error("leastCommonAncestor requires at least one argument");
+      if (forms.size() == 1) return *forms.begin();
+
+      std::initializer_list<Form>::const_iterator iter = forms.begin();
+      Form ret = *iter;
+      while (++iter != forms.end()) {
+        ret = leastCommonAncestor(ret, *iter);
+      }
+      return ret;
+    }
+
+    static InputOutputRelation composeWith(const InputOutputRelation &g,
+                                           const InputOutputRelation &f) { // composition of functions y = g(f(x))
+      return InputOutputRelation(leastCommonAncestor(g.form, f.form));
+    }
+
+    /*
+    static InputOutputRelation linearCombination(const InputOutputRelation& g, const InputOutputRelation& f) {
+      return InputOutputRelation(leastCommonAncestor({g.form,f.form,LINEAR}));
+    }*/
+    static InputOutputRelation combine(const InputOutputRelation &a,
+                                       const InputOutputRelation &b) { // vertical concatenation
+      return InputOutputRelation(leastCommonAncestor(a.form, b.form));
+    }
+
+    static InputOutputRelation combine(std::initializer_list<InputOutputRelation> args) {
+      if (args.size() < 1) throw std::runtime_error("combine requires at least one argument");
+      if (args.size() == 1) return *args.begin();
+
+      std::initializer_list<InputOutputRelation>::const_iterator iter = args.begin();
+      InputOutputRelation ret = *iter;
+      while (++iter != args.end()) {
+        ret = combine(ret, *iter);
+      }
+      return ret;
+    }
+
+  private:
+    static Form derivesFrom(const Form &f) { // capture the inheritance relationships (without resorting to using types)
+      switch (f) {
+        case DIFFERENTIABLE :
+          return ARBITRARY;
+        case POLYNOMIAL:
+          return DIFFERENTIABLE;
+        case AFFINE:
+          return POLYNOMIAL;
+        case LINEAR:
+          return AFFINE;
+        case CONSTANT:
+          return AFFINE;
+        case ZERO:
+          return LINEAR; // note: really want multiple inheritance here, since it's also constant
+        default:
+          return ARBITRARY;
+      }
+    }
+  };
+
+  class Function {
+  public:
+    Function() : relation(InputOutputRelation::ARBITRARY) {};
+    Function(InputOutputRelation::Form f) : relation(f) {};
+    virtual ~Function() {};
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const = 0;
+    virtual const InputOutputRelation& getInputOutputRelation() const { return relation; };
+
+  protected:
+    InputOutputRelation relation;
```

Comment:
added a comment to what I think the implementation might look like for that case.